File: Python/1774_Closest_Dessert_Cost.py
class Solution:
    result = float('inf')
    def closestCost(self, baseCosts: List[int], toppingCosts: List[int], target: int) -> int:
        N, M = len(baseCosts), len(toppingCosts)
        def dfs(idx, cost):
            if idx < M:
                dfs(idx + 1, cost)
                dfs(idx + 1, cost + toppingCosts[idx])
                dfs(idx + 1, cost + 2 * toppingCosts[idx])
            else:
                for base in baseCosts:
                    check(base + cost)
        
        def check(total):
            if abs(total - target) < abs(self.result - target):
                self.result = total
            elif abs(total - target) == abs(self.result - target) and total < self.result:
                self.result = total
        
        dfs(0, 0)
        return self.result

File: CPlusPlus/1774_Closest_Dessert_Cost.cpp
class Solution {
public:
    void check(int total, int target, int& result) {
        if (abs(total - target) < abs(result - target)) {
            result = total;
        } else if (abs(total - target) == abs(result - target) && total < result) {
            result = total;
        }
    }

    void dfs(vector<int>& baseCosts, vector<int>& toppingCosts, int target, int idx, int cost, int& result) {
        if (idx < toppingCosts.size()) {
            dfs(baseCosts, toppingCosts, target, idx + 1, cost, result);
            dfs(baseCosts, toppingCosts, target, idx + 1, cost + toppingCosts[idx], result);
            dfs(baseCosts, toppingCosts, target, idx + 1, cost + 2 * toppingCosts[idx], result);
        } else {
            for (auto& base : baseCosts) {
                check(base + cost, target, result);
            }
        }
    }
    
    int closestCost(vector<int>& baseCosts, vector<int>& toppingCosts, int target) {
        int result = INT_MAX;
        dfs(baseCosts, toppingCosts, target, 0, 0, result);
        return result;
    }
};
